lib/PhasarLLVM/ControlFlow/Resolver/RTAResolver.cpp

```diff
@@ -95,61 +92,23 @@ auto RTAResolver::resolveVirtualCall(const llvm::CallBase *CallSite)
 std::string RTAResolver::str() const { return "RTA"; }
 
 /// More or less copied from GeneralStatisticsAnalysis
-void RTAResolver::resolveAllocatedStructTypes() {
-  if (!AllocatedStructTypes.empty()) {
+void RTAResolver::resolveAllocatedCompositeTypes() {
+  if (!AllocatedCompositeTypes.empty()) {
     return;
   }
 
-  llvm::DenseSet<const llvm::StructType *> AllocatedStructTypes;
-
-  for (const auto *Fun : IRDB->getAllFunctions()) {
-    for (const auto &Inst : llvm::instructions(Fun)) {
-      if (const auto *Alloca = llvm::dyn_cast<llvm::AllocaInst>(&Inst)) {
-        if (const auto *StructTy =
-                llvm::dyn_cast<llvm::StructType>(Alloca->getAllocatedType())) {
-          AllocatedStructTypes.insert(StructTy);
-        }
-      } else if (const auto *CallSite = llvm::dyn_cast<llvm::CallBase>(&Inst);
-                 CallSite && CallSite->getCalledFunction()) {
-        // check if an instance of a user-defined type is allocated on the
-        // heap
-
-        if (!isHeapAllocatingFunction(CallSite->getCalledFunction())) {
-          continue;
-        }
-        /// TODO: Does this iteration over the users make sense?
-        /// After LLVM 15 we will probably not be able to access the
-        /// PointerElementType anyway...
-        for (const auto *User : Inst.users()) {
-          const auto *Cast = llvm::dyn_cast<llvm::BitCastInst>(User);
-          if (!Cast || Cast->getDestTy()->isOpaquePointerTy() ||
-              !Cast->getDestTy()
-                   ->getNonOpaquePointerElementType()
-                   ->isStructTy()) {
-            continue;
-          }
-          // finally check for ctor call
-          for (const auto *User : Cast->users()) {
-            if (const auto *CTor = llvm::dyn_cast<llvm::CallBase>(User)) {
-              // potential call to the structures ctor
-              if (CTor->getCalledFunction() &&
-                  getNthFunctionArgument(CTor->getCalledFunction(), 0)
-                          ->getType() == Cast->getDestTy() &&
-                  !Cast->getDestTy()->isOpaquePointerTy()) {
-                if (const auto *StructTy = llvm::dyn_cast<llvm::StructType>(
-                        Cast->getDestTy()->getNonOpaquePointerElementType())) {
-                  AllocatedStructTypes.insert(StructTy);
-                }
-              }
-            }
-          }
-        }
-      }
+  llvm::DebugInfoFinder DIF;
+  DIF.processModule(*IRDB->getModule());
+  llvm::DenseSet<const llvm::DICompositeType *> AllocatedCompositeTypes;
```

Comment:
This probably does not need to be deduplicated; we may get rid of the DenseSet here and directly push into the vector